src/bin/replicator.rs

```diff
@@ -0,0 +1,139 @@
+#[macro_use]
+extern crate clap;
+extern crate getopts;
+extern crate serde_json;
+#[macro_use]
+extern crate solana;
+
+use clap::{App, Arg};
+use solana::blob_fetch_stage::BlobFetchStage;
+use solana::crdt::{Crdt, Node, NodeInfo};
+use solana::fullnode::Config;
+use solana::logger;
+use solana::ncp::Ncp;
+use solana::service::Service;
+use solana::signature::{Keypair, KeypairUtil};
+use solana::store_ledger_stage::StoreLedgerStage;
+use solana::window;
+use solana::window_service::window_service;
+use std::fs::File;
+use std::net::UdpSocket;
+use std::net::{Ipv4Addr, SocketAddr};
+use std::process::exit;
+use std::sync::atomic::AtomicBool;
+use std::sync::mpsc::channel;
+use std::sync::{Arc, RwLock};
+
+fn main() {
+    logger::setup();
+    let matches = App::new("replicator")
+        .version(crate_version!())
+        .arg(
+            Arg::with_name("identity")
+                .short("i")
+                .long("identity")
+                .value_name("PATH")
+                .takes_value(true)
+                .help("Run with the identity found in FILE"),
+        ).arg(
+            Arg::with_name("network")
+                .short("n")
+                .long("network")
+                .value_name("HOST:PORT")
+                .takes_value(true)
+                .help("Rendezvous with the network at this gossip entry point"),
+        ).arg(
+            Arg::with_name("ledger")
+                .short("l")
+                .long("ledger")
+                .value_name("DIR")
+                .takes_value(true)
+                .required(true)
+                .help("use DIR as persistent ledger location"),
+        ).get_matches();
+
+    let ledger_path = matches.value_of("ledger");
+
+    let (keypair, ncp) = if let Some(i) = matches.value_of("identity") {
+        let path = i.to_string();
+        if let Ok(file) = File::open(path.clone()) {
+            let parse: serde_json::Result<Config> = serde_json::from_reader(file);
+            if let Ok(data) = parse {
+                (data.keypair(), data.node_info.contact_info.ncp)
+            } else {
+                eprintln!("failed to parse {}", path);
+                exit(1);
+            }
+        } else {
+            eprintln!("failed to read {}", path);
+            exit(1);
+        }
+    } else {
+        (Keypair::new(), socketaddr!([127, 0, 0, 1], 8700))
+    };
+
+    let node = Node::new_with_external_ip(keypair.pubkey(), &ncp);
+
+    println!(
+        "replicating the data with keypair: {:?} ncp:{:?}",
+        keypair.pubkey(),
+        ncp
+    );
+    println!("my node: {:?}", node);
+
+    let exit = Arc::new(AtomicBool::new(false));
```

Comment:
Can you move just about everything from here on down to a separate `src/replicator.rs` module? And add some tests for it.